### ISSUE TITLE
Unable to find a valid endpoint for function mail

### ISSUE THREAD

**Issue by christianloosli** [firebase/firebase-tools#4792]
### [REQUIRED] Environment info

**firebase-tools:** 11.4.0

**Platform:** macOS and Github Actions

### [REQUIRED] Steps to reproduce

`firebase deploy --only hosting:sometarget`  or 
```
      - uses: FirebaseExtended/action-hosting-deploy@v0
        with:
          repoToken: '${{ secrets.GITHUB_TOKEN }}'
          firebaseServiceAccount: '${{ secrets.FIREBASE_SERVICE_ACCOUNT_DEV }}'
          projectId: xxxxxx
          channelID: live
          target: ${{ env.TARGET }}
        env:
          FIREBASE_CLI_PREVIEWS: hostingchannels
```
### [REQUIRED] Expected behavior

Should deploy successfully, because last deployment succeeded on 17 Jun without code changes.

```
***
  details_url: 'https://xxx.web.app/',
  conclusion: 'success',
  output: ***
    title: 'Production deploy succeeded',
    summary: '[xxx.web.app](https://xxx.web.app/)'
  ***
***
```

### [REQUIRED] Actual behavior

Deployment fails with the following error message:
````
  ***
    "status": "error",
    "error": "Unable to find a valid endpoint for function mail"
  ***
```
<!-- Run the command with --debug flag, and include the logs below. -->

**Comment by onatdesigndev** [comment]
I encountered the same issue after updating firebase to the latest version which is 11.4.0

**Comment by christianloosli** [comment]
ok great, I will try that. Unfortunately, this does not solve our problem for our CD pipeline. The version in FirebaseExtended/action-hosting-deploy@v0 can't be overwritten, or can we?

**Comment by muryoh** [comment]
> I encountered the same issue after updating firebase to the latest version which is 11.4.0
> 
> UPDATE: I had to uninstall the latest version of firebase and install the previous version which is 11.3.0 IT WORKED!!
> 
> I hope it's the same case for you so you can also resolve this for your project.
> 
> By the way, these are the commands for uninstalling and installing an older version of firebase: npm uninstall -g firebase-tools npm install -g firebase-tools@11.3.0

I confirm downgrading worked on my end, thanks

**Comment by JGSolutions** [comment]
same issue. will downgrade now and will try again

**Comment by bkendall** [comment]
Hi! Thank you for raising. I'm certain I know where this issue will stem from, but just have to do some debugging.

**Comment by bkendall** [comment]
@christianloosli and other folks here, could you let me know something while I look?

1. Confirm for me that you're deploying a Hosting site with a `rewrite` configured to a function.
2. Is that function in the same codebase from which you are deploying?
3. If the function is not in the same codebase, is the function deployed in your project already?

**Comment by jnstahl** [comment]
Same problem here. found one rewrite to a function that was no longer there and removed that rewrite.
But now it fails with another function that actually is deployed from the same codebase.
some functions work however.

**Comment by bkendall** [comment]
@jnstahl I think that is what I've narrowed in on myself - it's a bit of a change of behavior I think I'm going to flip back a bit and make it print a warning instead.

Can you give me a minimal repro case for the function that's failing but in the same codebase? That'd be extremely useful.

**Comment by alexandru-g** [comment]
Same issue here. What I can see from debugging is that it queries for v1 functions (which are indeed returned) and then it queries for v2 functions. This response is empty (as we're using v1) and then the 'Error: Unable to find a valid endpoint for function' error is thrown.

Hopefully this is not a weird way we are being forced into v2 functions 😬

**Comment by bkendall** [comment]
@alexandru-g definitely not forcing you to v2 😛 . Is your function a `v1` function that is already deployed and the CLI doesn't seem to be catching it? That would be a good lead for us to track down. If you can give us some specifics about the situation it would help us debug correctly!

**Comment by cie-cloosli** [comment]
> @christianloosli and other folks here, could you let me know something while I look?
> 
> 1. Confirm for me that you're deploying a Hosting site with a `rewrite` configured to a function.
> 2. Is that function in the same codebase from which you are deploying?
> 3. If the function is not in the same codebase, is the function deployed in your project already?

Hi @bkendall, downgrade to v11.3.0 solved the problem for local deployment.
1. yes we have a `rewrite` configuration in place `{ "source": "/mail", "function": "mail" },`
2. No it is not in the same codebase
3. Not in this environment, but before it was not a problem

**Comment by alexandru-g** [comment]
> @alexandru-g definitely not forcing you to v2 😛 . Is your function a `v1` function that is already deployed and the CLI doesn't seem to be catching it? That would be a good lead for us to track down. If you can give us some specifics about the situation it would help us debug correctly!

Hey! Yes, the v1 function was already deployed for a while and everything was working correctly. I've tried deleting and redeploying. What's important is that the issue appeared all of the sudden, with no changes on our part in regards to the functions or hosting rewrites (I think we can exclude a bug on our side). It must have been an update on the tools (inside GitHub Actions as well).

**Comment by alexandru-g** [comment]
I have no idea how the code looks like, but seeing that it's first querying for V1 functions and then for V2 functions (which is empty), it must replace a collection instead of appending to it somewhere :)

**Comment by alexandru-g** [comment]
@bkendall I found the issue and validated the fix:
```
diff --git a/src/deploy/hosting/convertConfig.ts b/src/deploy/hosting/convertConfig.ts
index 602b8436..3b0f225d 100644
--- a/src/deploy/hosting/convertConfig.ts
+++ b/src/deploy/hosting/convertConfig.ts
@@ -1,6 +1,6 @@
 import { FirebaseError } from "../../error";
 import { HostingConfig, HostingRewrites, HostingHeaders } from "../../firebaseConfig";
-import { existingBackend, allEndpoints, isHttpsTriggered } from "../functions/backend";
+import { existingBackend, allEndpoints, isHttpsTriggered, isCallableTriggered } from "../functions/backend";
 import { Payload } from "./args";
 import * as backend from "../functions/backend";
 import { Context } from "../functions/args";
@@ -88,7 +88,7 @@ export async function convertConfig(
 
     if (matchingBackends.length === 1) {
       const endpoint = matchingBackends[0];
-      if (endpoint && isHttpsTriggered(endpoint)) {
+      if (endpoint && (isHttpsTriggered(endpoint) || isCallableTriggered(endpoint))) {
         return endpoint;
       }
     }
```

It was previously only calling ```isHttpsTriggered(endpoint)``` which I think it's an incorrect assumption, as my function is indeed a callable and not normal http function.

**Comment by bkendall** [comment]
@alexandru-g so to be clear, you have a [*callable function*](https://firebase.google.com/docs/functions/callable) that you are *rewriting* to in your Hosting config?

That's a tricky situation. According to the docs (and to my knowledge), rewriting to a callable function isn't supported. It _may work_ because the underlying implementation is a HTTP call, but Hosting rewrites are to HTTP functions (`functions.https.onRequest`), not callable functions (`functions.https.onCall`).

I'll run this by the team and see what they say.

@cie-cloosli I think I have a fix for you in #4799. We were throwing an error if a function wasn't found in the codebase or project which is a breaking behavior from what we were doing before. I've reverted that change and am shipping a new version of the CLI soon with the fix (`v11.4.1` - coming soon)

**Comment by jnstahl** [comment]
Indeed looking at the code it is the rewrites on the  onCall functions that cause the problem. But so far this has worked all correctly, so why should it not be supported? AFAIK this is the only way to keep callable functions in the same domain and to avoid CORS issues

**Comment by alexandru-g** [comment]
@bkendall yes, it's a callable function.

I would say it's 100% supported as it was working perfectly before. The idea is that we use a custom hostname in the Cloud Functions API depending on the compiled flavor (but that's not important) and that needs the rewrite to work. Long story short, we'd like all our functions to be accessible using a single hostname - our domain (hence the rewrite).

**Comment by bkendall** [comment]
Alright, discussions had, decisions made. We're gonna make sure callable functions are candidates for rewrites. PR'll come sone and hopefully released later today. Thanks for y'all's patience and info as we got this debugged!

**Comment by taeold** [comment]
Catching up on the discussion - thanks for clear communication @jnstahl @alexandru-g!

One thing I'm curious about - callable functions are designed to be called by Firebase client SDKs, so I didn't anticipate them being used as Hosting rewrite targets (though we don't try very hard to hide that it's just HTTP function behind the scenes: [public doc](https://firebase.google.com/docs/functions/callable-reference)). I don't _think_ Firebase Client SDK will ever attempt to use the Firebase Hosting url to call these functions.

Callable functions do have some nice properties - like automatically detecting and validating auth and app check tokens. Wondering what about made you folks choose callable function over just plain http functions?

**Comment by jnstahl** [comment]
exactly this property. I have some single-page web apps which renders content from the firestore, but relies on the callable functions to perform certain privileged operations. The callable function will use the auth token to establish the user ID and will then determine if this particular user is authorized to perform the privileged transaction.

**Comment by alexandru-g** [comment]
@taeold you sound like you have access to our repo :). We thoroughly enjoy having auth and app check handled by Firebase magic!
Another 2 reasons for us are that you mess a lot less with CORS and callable functions play way more nicely with mobile function SDKs (at least in our experience).

**Comment by bkendall** [comment]
Alright, I've confirmed that `v11.4.1` should actually "fix" the deploys that were failing to find the callable functions. It fixes it by simply ignoring the fact that the CLI can't find the function (because it only looks for https). I'll make a small PR that allows callable function rewrites as well which will get rid of the warning message that's printed, but y'all _should_ be unblocked on the latest release.

Let me know if you're still seeing issues.

**Comment by jnstahl** [comment]
> Alright, I've confirmed that `v11.4.1` should actually "fix" the deploys that were failing to find the callable functions. It fixes it by simply ignoring the fact that the CLI can't find the function (because it only looks for https). I'll make a small PR that allows callable function rewrites as well which will get rid of the warning message that's printed, but y'all _should_ be unblocked on the latest release.
> 
> Let me know if you're still seeing issues.

OK the deploy now works again, but I noticed the following warning:

 hosting[swimtimer1]: Unable to find a valid endpoint for function `webSignUp`, but still including it in the config

now 'webSignUp' is actually a function that no longer exists (neither https nor callable) - The other functions do not produce a warning regardless of call or request.

Not sure if this is a bug or if it was intended to be this way.

**Comment by bkendall** [comment]
> now 'webSignUp' is actually a function that no longer exists (neither https nor callable) - The other functions do not produce a warning regardless of call or request.

Yup, this sounds like it's working as intended. If the CLI is unable to find the rewrite target function (either HTTPS or callable) in either the local codebase _or_ the live project, it'll print that warning. Since it's specified in your `firebase.json` configuration, we're just going to print a warning for now rather than error out - you can choose to leave it or remove it if it's a function that really is gone!

**Comment by bkendall** [comment]
Sounds like this is resolved now - if there's more specific cases that are broken, feel free to open a new issue!

**Comment by VISHNUONE8** [comment]
Hi, @bkendall i am using gen2 and 14.0.0 firebase-tools and still facing the issue not a valid endpoint 
can you help me with this...
